server: notify tenant servers of metadata changes

knz opened the pull request:
Informs (towards fixing) #98431.
Fixes #105721.
Epic: CRDB-26691

See the individual commits for details.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105441)
<!-- Reviewable:end -->

knz commented:
Note to self:

- complete the connector-side of the change
- rename all the things

yuzefovich reviewed (commented):
Seems good to me. As always, thanks for great commit messages.

> Reviewed 4 of 4 files at r1, 12 of 12 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-:-NYslJpKBhIO6Q0w8UxV:bwkeuip)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsa5ut5Hm3vSU-orWz):*
There is another `log.Fatal` call in `pkg/multitenant` directory - in `decoder.go`, should that be fixed too?
___
*[`-- commits` line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsgYio4AqYyAVljTqE):*
nit: `s/medata/metadata/`.
___
*[`-- commits` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsh_wzESFgX9VN3sG1):*
nit: I think you meant `pkg/server/tenantsettingswatcher/watcher.go`.
___
*[`-- commits` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsjGRB-3sQoU541O61):*
This seems like an opportunity to extract a helper to avoid code duplication, and I'm assuming you considered it but decided not to, what's the rationale?
___
*[`pkg/kv/kvpb/api.proto` line 3131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYshI3Q0GTIB1nbrI6x:-NYshI3Q0GTIB1nbrI6y:bh2ozff) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5bdedcddb5cd978cdc89901a4c7bcba8ee0b069/pkg/kv/kvpb/api.proto#L3131)):*
> ```proto
>   // Note that this field is advisory: the server may know of a more
>   // recent (and different) set of capabilities, and server-side
>   // capabilities check always prevail.
> ```

nit: `s/prevail/prevails/`.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/BUILD.bazel` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsbkPZCs6dGmRamRk5:-NYsbkPZCs6dGmRamRk6:bxnkuo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/BUILD.bazel#L4)):*
> ```Starlark
> 
> go_library(
>     name = "tenantcapabilitiestestutils",
> ```

nit: we could add `testonly = 1,` while we're here.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/testutils.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsfZ4XBqh-c_AJRqWU:-NYsfZ4XBqh-c_AJRqWV:b5ibkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/testutils.go#L58)):*
> ```Go
> ) (
> 	name *roachpb.TenantName,
> 	serviceMode mtinfopb.TenantServiceMode,
> ```

nit: should we switch the order of `serviceMode` and `dataState` here and in `watcher_test.go` to match the order of columns in the system table as well as in the `Entry` struct? Up to you.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYscarO9Z-5hPA7wC6G:-NYscarP0f60jrvO_ECJ:bwdm92a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L92)):*
> ```Go
> 
> 	var name roachpb.TenantName
> 	if len(datums) > 3 {
> ```

Why do we need length checks on `datums` here? My reading of the code says that `Decoder.Decode` should return a datum for each column in the table, and `system.tenants` has 6 columns.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsdDSA2FcUY0JRqqVz:-NYsdDSA2FcUY0JRqqW-:beuh8gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L102)):*
> ```Go
> 		if i := datums[4]; i != tree.DNull {
> 			rawDataState := tree.MustBeDInt(i)
> 			if rawDataState >= 0 && rawDataState <= tree.DInt(mtinfopb.MaxDataState) {
> ```

It's unexpected when `rawDataState` is outside of `[0, mtinfopb.MaxDataState)` range, right? Should we return an error or at least log a warning? The only valid scenario when this can happen I can think of is when a user manually inserts a row with these values into the system table, are there other cases?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYseFGG5H_Y6RVh69vV:-NYseFGG5H_Y6RVh69vW:b-w2fryt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L129)):*
> ```Go
> }
> 
> // capabilityEntrySize is an estimate for a (tenantID, capability) pair that the
> ```

nit: the comment is now stale.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsil6M0PmSEW0nbQzZ:-NYsil6M0PmSEW0nbQz_:behz6c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5bdedcddb5cd978cdc89901a4c7bcba8ee0b069/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L158)):*
> ```Go
> 	id roachpb.TenantID,
> ) (*tenantcapabilitiespb.TenantCapabilities, bool) {
> 	cp := w.getInternal(id)
> ```

This call to `getInternal` increases the mutex contention in case when the entry is not found (previously we only had a single RLock, but now we'll have RLock followed by RWLock). Is this concerning?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsa5ut5Hm3vSU-orWz):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

There is another `log.Fatal` call in `pkg/multitenant` directory - in `decoder.go`, should that be fixed too?
</blockquote></details>

Good idea. Done.
___
*[`-- commits` line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsgYio4AqYyAVljTqE):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/medata/metadata/`.
</blockquote></details>

Fixed.
___
*[`-- commits` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsh_wzESFgX9VN3sG1):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think you meant `pkg/server/tenantsettingswatcher/watcher.go`.
</blockquote></details>

Indeed. Fixed.
___
*[`-- commits` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsjGRB-3sQoU541O61):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This seems like an opportunity to extract a helper to avoid code duplication, and I'm assuming you considered it but decided not to, what's the rationale?
</blockquote></details>

Alas the sharing of concepts here is in the programming pattern -- the repeated code is interleaved with the rest. It's not yet clear (to me) how to extract that abstraction into a library.
___
*[`pkg/kv/kvpb/api.proto` line 3131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYshI3Q0GTIB1nbrI6x:-NYwrE-R4cf0XeF3wZZh:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5bdedcddb5cd978cdc89901a4c7bcba8ee0b069/pkg/kv/kvpb/api.proto#L3131)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/prevail/prevails/`.
</blockquote></details>

Fixed.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/BUILD.bazel` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsbkPZCs6dGmRamRk5:-NYwrRLK8tC3zB9IIhIz:b-k6yvij) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/BUILD.bazel#L4)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we could add `testonly = 1,` while we're here.
</blockquote></details>

Nice TIL.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/testutils.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsfZ4XBqh-c_AJRqWU:-NYwrkjd0qPacwAh6Rve:bigto6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils/testutils.go#L58)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we switch the order of `serviceMode` and `dataState` here and in `watcher_test.go` to match the order of columns in the system table as well as in the `Entry` struct? Up to you.
</blockquote></details>

Sure why not.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYscarO9Z-5hPA7wC6G:-NYwukfZ5F7yTxtqh5Eu:b-4i1cda) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L92)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why do we need length checks on `datums` here? My reading of the code says that `Decoder.Decode` should return a datum for each column in the table, and `system.tenants` has 6 columns.
</blockquote></details>

Clarified in comment. This needs to work also with servers that haven't run the V23_1TenantNamesStateAndServiceMode upgrade migration yet, where these columns are added.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsdDSA2FcUY0JRqqVz:-NYwvT3051xvFelfhbIh:bitxlh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L102)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It's unexpected when `rawDataState` is outside of `[0, mtinfopb.MaxDataState)` range, right? Should we return an error or at least log a warning? The only valid scenario when this can happen I can think of is when a user manually inserts a row with these values into the system table, are there other cases?
</blockquote></details>

Let's do a warning. Indeed, your understanding is right this is defensive programming.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYseFGG5H_Y6RVh69vV:-NYwsIh-1wvFjOqHuZBW:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L129)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the comment is now stale.
</blockquote></details>

Fixed.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYsil6M0PmSEW0nbQzZ:-NYwvpU3BUIsx900qmbE:bxiggr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5bdedcddb5cd978cdc89901a4c7bcba8ee0b069/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L158)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This call to `getInternal` increases the mutex contention in case when the entry is not found (previously we only had a single RLock, but now we'll have RLock followed by RWLock). Is this concerning?
</blockquote></details>

The rate at which capabilities are updated is presumably very low (it's supposed to be a manual operation).

Also FYI this is the same pattern used in `tenantsettingswatcher`, see `(s *overridesStore) GetTenantOverrides()` in `overrides_store.go`.


<!-- Sent from Reviewable.io -->

knz committed:
kvtenant: receive the tenant metadata in the connector

The previous commit made the server send the metadata through the
TenantSettingEvent streaming RPC.

This commit makes the tenant connector receive and store the metadata.

Like in the previous commits, we acknowledge that the whole
infrastructure has become more generic than just "cluster settings"
and so the components involved are now misnamed. The appropriate
renaming will be performed in a later stage.

Release note: None

yuzefovich reviewed (commented):
> Reviewed 21 of 21 files at r4, 12 of 12 files at r5, 15 of 16 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`pkg/kv/kvpb/api.proto` line 3099 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYy_-zE98BVB3Y7wZEa:-NYy_-zE98BVB3Y7wZEb:b-j791xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/52d9c5c882d658b6a0297a2b5dd4b14a6a9daf3a/pkg/kv/kvpb/api.proto#L3099)):*
> ```proto
>   enum EventType {
>     // The event is about an update to cluster setting overrides.
>     // This must be zero for backward-compatibility with pre-v23.1
> ```

nit: should this be `pre-v23.1`? Or are you intending to backport this to 23.1?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYscarO9Z-5hPA7wC6G:-NYy_n1p27241_QvdxQ1:b50gq43) ([raw file](https://github.com/cockroachdb/cockroach/blob/c706d5fb2963154dabe991da3ce937b6cf4b5bef/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L92)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Clarified in comment. This needs to work also with servers that haven't run the V23_1TenantNamesStateAndServiceMode upgrade migration yet, where these columns are added.
</blockquote></details>

Ack, makes sense.


<!-- Sent from Reviewable.io -->

knz commented:
@yuzefovich could you have another look? Before I embark on the project to rename all the things.

yuzefovich reviewed (commented):
LGTM, I only have some minor questions / comments.

> Reviewed 23 of 23 files at r7, 1 of 1 files at r8, 7 of 7 files at r9, 13 of 13 files at r10, 10 of 10 files at r11, 4 of 4 files at r12, 6 of 6 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-:-NZ40Lk69NrQwEQNjOU9:b-xjyd2x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`-- commits` line 102 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3vDXGCS5qfFgvHuWU):*
The protocol seems sound to me, but I feel like this deserves a mixed-version unit test, is it possible to add one?
___
*[`pkg/kv/kvclient/kvtenant/BUILD.bazel` line 25 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3zYKB2OINIqVPKLLR:-NZ3zYKB2OINIqVPKLLS:b-p9pien) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e3c10666014da4130e6b1b1c8b87200a71ae9ce/pkg/kv/kvclient/kvtenant/BUILD.bazel#L25)):*
> ```Starlark
>         "//pkg/multitenant/mtinfopb",
>         "//pkg/multitenant/tenantcapabilities",
>         "//pkg/multitenant/tenantcapabilities/tenantcapabilitiespb",
> ```

When renaming things, WDYT about making `tenantcapabilities` shorter, perhaps `tenantcaps`? Or I guess with the renaming it'll be something like `tenantinfo` or `tenantmetadata`?
___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 89 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3zzeT8xpIXP8NeoJy:-NZ3zzeT8xpIXP8NeoJz:b-sca5tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e3c10666014da4130e6b1b1c8b87200a71ae9ce/pkg/kv/kvclient/kvtenant/setting_overrides.go#L89)):*
> ```Go
> 				// Signal that startup is complete once we have enough events
> 				// to start. Note: we do not connect this condition to
> 				// receiving the tenant medata (via processMetadataEvent) for
> ```

nit: `s/medata/metadata/`.
___
*[`pkg/kv/kvpb/api.proto` line 3139 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3vgxiDzCNg5CZNsUF:-NZ3vgxiDzCNg5CZNsUG:b-ynyuhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90954803f1bae1f4e2b615b68006a6625b410ec/pkg/kv/kvpb/api.proto#L3139)):*
> ```proto
> 
>   // The type of event. For backward-compatibility with servers that
>   // do not check the event_type field, all events of other types
> ```

nit: accessing `event_type` field should be allowed once the cluster has been upgraded to 23.2, right? Should we leave a TODO or adjust the comment to mention this?
___
*[`pkg/kv/kvpb/api.proto` line 3202 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3vyeA95HkWKehZN-1:-NZ3vyeA95HkWKehZN-2:bqva9fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90954803f1bae1f4e2b615b68006a6625b410ec/pkg/kv/kvpb/api.proto#L3202)):*
> ```proto
> 
>   // ServiceMode is the tenant's current service mode.
>   // TODO(knz): This should really be casted to go type mtinfopb.TenantServicemode but we
> ```

nit: `s/Servicemode/ServiceMode/`.
___
*[`pkg/server/node.go` line 2163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3yTW20MIAKTkDrwob:-NZ3yTW20MIAKTkDrwoc:bpg7f7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90954803f1bae1f4e2b615b68006a6625b410ec/pkg/server/node.go#L2163)):*
> ```Go
> 	// - For compatibility with older version tenant clients,
> 	//   all non-setting event types must fake being a no-op
> 	//   setting event (see the docstring on event_type below).
> ```

nit: there is no `event_type` below.
___
*[`pkg/kv/kvclient/kvtenant/setting_overrides_test.go` line 241 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ4021u9_BMzsV9bDEe:-NZ4021u9_BMzsV9bDEf:b-qegqmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e3c10666014da4130e6b1b1c8b87200a71ae9ce/pkg/kv/kvclient/kvtenant/setting_overrides_test.go#L241)):*
> ```Go
> 	}
> 	eventCh <- ev
> 	select {
> ```

What is the rationale for blocking for 10ms after each event? I'd assume we could just send as many as we want and only block once, after all have been sent, for up to 10s.


<!-- Sent from Reviewable.io -->

knz committed:
kvtenant: define a cross-version metadata protocol test

This commit introduces a new test to check that old-version tenant
clients can operate with a new-version KV server, and vice-versa.

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 102 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3vDXGCS5qfFgvHuWU):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The protocol seems sound to me, but I feel like this deserves a mixed-version unit test, is it possible to add one?
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvtenant/BUILD.bazel` line 25 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3zYKB2OINIqVPKLLR:-NZkuphcCRd-Q6h2u_D-:b-454reo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e3c10666014da4130e6b1b1c8b87200a71ae9ce/pkg/kv/kvclient/kvtenant/BUILD.bazel#L25)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

When renaming things, WDYT about making `tenantcapabilities` shorter, perhaps `tenantcaps`? Or I guess with the renaming it'll be something like `tenantinfo` or `tenantmetadata`?
</blockquote></details>

Sounds good. I'll do this in a separate PR.
___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 89 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3zzeT8xpIXP8NeoJy:-NZkv-ZW0pGeUS13Udlw:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e3c10666014da4130e6b1b1c8b87200a71ae9ce/pkg/kv/kvclient/kvtenant/setting_overrides.go#L89)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/medata/metadata/`.
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvpb/api.proto` line 3099 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NYy_-zE98BVB3Y7wZEa:-NZkv0qC7ZdyAvaEOXc8:byy0l41) ([raw file](https://github.com/cockroachdb/cockroach/blob/52d9c5c882d658b6a0297a2b5dd4b14a6a9daf3a/pkg/kv/kvpb/api.proto#L3099)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should this be `pre-v23.1`? Or are you intending to backport this to 23.1?
</blockquote></details>

yes this should remain backportable.
___
*[`pkg/kv/kvpb/api.proto` line 3139 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3vgxiDzCNg5CZNsUF:-NZkv3UJ7GsLGFDaSVdY:b1h604i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90954803f1bae1f4e2b615b68006a6625b410ec/pkg/kv/kvpb/api.proto#L3139)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: accessing `event_type` field should be allowed once the cluster has been upgraded to 23.2, right? Should we leave a TODO or adjust the comment to mention this?
</blockquote></details>

I don't understand your question?
___
*[`pkg/kv/kvpb/api.proto` line 3202 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3vyeA95HkWKehZN-1:-NZkvMsW0K9C3bNHhTWO:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90954803f1bae1f4e2b615b68006a6625b410ec/pkg/kv/kvpb/api.proto#L3202)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/Servicemode/ServiceMode/`.
</blockquote></details>

Fixed.
___
*[`pkg/server/node.go` line 2163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ3yTW20MIAKTkDrwob:-NZkvYye2ReH6JfzGtaz:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90954803f1bae1f4e2b615b68006a6625b410ec/pkg/server/node.go#L2163)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: there is no `event_type` below.
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvclient/kvtenant/setting_overrides_test.go` line 241 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZ4021u9_BMzsV9bDEe:-NZlN7zeENkg8bUv75M7:b-j7nt0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e3c10666014da4130e6b1b1c8b87200a71ae9ce/pkg/kv/kvclient/kvtenant/setting_overrides_test.go#L241)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

What is the rationale for blocking for 10ms after each event? I'd assume we could just send as many as we want and only block once, after all have been sent, for up to 10s.
</blockquote></details>

I was following the pattern already established in the rest of this file. The purpose for this pattern is that it ensures the test becomes flaky if the connector mistakenly reports readiness too early.
- 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 4 of 31 files at r14, 1 of 31 files at r21, 33 of 33 files at r28, 7 of 7 files at r29, 13 of 13 files at r30, 9 of 9 files at r31, 14 of 14 files at r32, 5 of 5 files at r33, 4 of 4 files at r34, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @renatolabs, @Santamaura, @smg260, and @stevendanna)

___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 74 at r32](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmjtwO7cPZ6-TJeXP0:-NZmjtwO7cPZ6-TJeXP1:b-sjwh46) ([raw file](https://github.com/cockroachdb/cockroach/blob/00213f1b63d1f0f1a2f6581c5415aa61d1e05f11/pkg/kv/kvclient/kvtenant/setting_overrides.go#L74)):*
> ```Go
> 				// Don't hammer the server with requests if there was an
> 				// actual error reported.
> 				select {
> ```

I don't quite understand this block. `runTenantSettingsSubscription` is called in a single spot in which we derive a child context that is canceled on quiesce, so why do we need first and second `case`s? Also why don't we just exit when the cancellation / quiesce happens? Why don't we exit when non-`io.EOF` error (other than `MissingRecordError`) is `Recv`ed?
___
*[`pkg/kv/kvpb/api.proto` line 3139 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmgkB43m15Ko5_3O01:-NZmgkB5CU429VoyXcBU:bebfaic) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a0b6bb432318aca0c3db14a5987fb4aaaa094a2/pkg/kv/kvpb/api.proto#L3139)):*
(Copy-pasting my previous comment to point at the correct line in the commit message):
> nit: accessing `event_type` field should be allowed once the cluster has been upgraded to 23.2, right? Should we leave a TODO or adjust the comment to mention this?

The second sentence in this paragraph to me implies that creators of `TenantSettingsEvent` must always do all 3 points to keep backwards-compatibility. However, IIUC this is only required until the cluster has been upgraded to 23.2, at which point `TenantSettingsEvent` can be created in any manner, contradicting any of the 3 points. Is that correct?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 130 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmg0_37A_UYIMnaDZZ:-NZmg0_47L40ZWkVJ-Bj:b9z4pvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9e10ae32cd1a6b71c73ed548855cbc38794617/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L130)):*
> ```Go
> 		err = errors.NewAssertionErrorWithWrappedErrf(err, "failed to decode row %v", ev.Key)
> 		var sv *settings.Values
> 		if d.st != nil {
> ```

nit: it seems like `nil` settings is only because of `TestingDecoderFn` which is only called in one spot, so I'd refactor the testing fn to pass valid settings object too, in order to remove this non-nil check.
___
*[`pkg/kv/kvclient/kvtenant/setting_overrides_test.go` line 317 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmmlpC5KN5T5POGmXD:-NZmmlpC5KN5T5POGmXE:bqelmo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/32eb14730edc2ae693412c01c751828a15e3a4a0/pkg/kv/kvclient/kvtenant/setting_overrides_test.go#L317)):*
> ```Go
> // can talk to an old-version connector.
> func TestCrossVersionMetadataSupport(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> ```

nit: maybe add `defer log.Scope(t).Close(t)`?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs, @Santamaura, @smg260, @stevendanna, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvtenant/setting_overrides.go` line 74 at r32](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmjtwO7cPZ6-TJeXP0:-NZzrSKFByP-20DnpoKr:bo7lteo) ([raw file](https://github.com/cockroachdb/cockroach/blob/00213f1b63d1f0f1a2f6581c5415aa61d1e05f11/pkg/kv/kvclient/kvtenant/setting_overrides.go#L74)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I don't quite understand this block. `runTenantSettingsSubscription` is called in a single spot in which we derive a child context that is canceled on quiesce, so why do we need first and second `case`s? Also why don't we just exit when the cancellation / quiesce happens? Why don't we exit when non-`io.EOF` error (other than `MissingRecordError`) is `Recv`ed?
</blockquote></details>

let's discuss on #106414
___
*[`pkg/kv/kvpb/api.proto` line 3139 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmgkB43m15Ko5_3O01:-NZzrWCp5Z0xdfnwUlBJ:bo7lteo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a0b6bb432318aca0c3db14a5987fb4aaaa094a2/pkg/kv/kvpb/api.proto#L3139)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

(Copy-pasting my previous comment to point at the correct line in the commit message):
> nit: accessing `event_type` field should be allowed once the cluster has been upgraded to 23.2, right? Should we leave a TODO or adjust the comment to mention this?

The second sentence in this paragraph to me implies that creators of `TenantSettingsEvent` must always do all 3 points to keep backwards-compatibility. However, IIUC this is only required until the cluster has been upgraded to 23.2, at which point `TenantSettingsEvent` can be created in any manner, contradicting any of the 3 points. Is that correct?
</blockquote></details>

let's discuss on #106414
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 130 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmg0_37A_UYIMnaDZZ:-NZzrdco7zvhA8Dot7VX:bo7lteo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9e10ae32cd1a6b71c73ed548855cbc38794617/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L130)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it seems like `nil` settings is only because of `TestingDecoderFn` which is only called in one spot, so I'd refactor the testing fn to pass valid settings object too, in order to remove this non-nil check.
</blockquote></details>

let's discuss on #106414


<!-- Sent from Reviewable.io -->

knz committed:
kvtenant: sprinkle some log scopes through tests

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105441)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs, @Santamaura, @smg260, @stevendanna, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvtenant/setting_overrides_test.go` line 317 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/105441#-NZmmlpC5KN5T5POGmXD:-N_-QC3o4K1qnLNHc5MF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32eb14730edc2ae693412c01c751828a15e3a4a0/pkg/kv/kvclient/kvtenant/setting_overrides_test.go#L317)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: maybe add `defer log.Scope(t).Close(t)`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz committed:
server: make tenant tests faster

This ensures that `WaitForTenantReadiness` doesn't have to wait for
the full closed timestamp interval.

Release note: None

stevendanna reviewed (approved):
Thanks for the productive review iteration here. I like where we've ended here.

knz commented on pkg/kv/kvclient/kvtenant/setting_overrides_test.go:
```diff
@@ -207,3 +222,169 @@ func expectSettings(t *testing.T, c *connector, exp string) <-chan struct{} {
+func expectMetadata(t *testing.T, c *connector, exp string) <-chan struct{} {
+	t.Helper()
+	info, updateCh := c.TenantInfo()
+	var buf strings.Builder
+	fmt.Fprintf(&buf, "tid=%v name=%q data=%v service=%v caps=%+v",
+		info.TenantID, info.Name, info.DataState, info.ServiceMode, info.TenantCapabilities)
+
+	str := buf.String()
```
Done.

knz commented:
TFYR!

bors r=stevendanna

knz commented:
CI is telling me somethin went wrong during the last rebase

bors r-

knz committed:
sctest: skip alter_table_add_primary_key_drop_rowid

Release note: None

knz commented:
bors r=stevendanna

knz commented:
bors r-

knz commented:
bors r=yuzefovich,stevendanna
